File: src/mhashim6/game1024/GameImpl.java
package mhashim6.game1024;

import mhashim6.game1024.Exceptions.GameOverException;

public class GameImpl implements Game {

	private int maxValue;
	private GridOfTiles gridOfTiles;

	private int movesCount;

	private boolean isVictory;

	private boolean isStarted;
	private boolean isGameOver;

	public GameImpl(int power, int gridSize) {
		maxValue = (int) Math.pow(2, power); //11 -> 2048

		gridOfTiles = new GridOfTiles(gridSize);
	}

	@Override
	public Tile[][] start() {
		if (isStarted)
			throw new RuntimeException("The game is already on");
		isStarted = true;

		gridOfTiles.newTile();
		gridOfTiles.newTile();

		return gridOfTiles.get();
	}

	@Override
	public Tile[][] swipeUp() {
		swipe(Direction.UP);
		return gridOfTiles.get();
	}

	@Override
	public Tile[][] swipeDown() {
		swipe(Direction.DOWN);
		return gridOfTiles.get();
	}

	@Override
	public Tile[][] swipeRight() {
		swipe(Direction.RIGHT);
		return gridOfTiles.get();
	}

	@Override
	public Tile[][] swipeLeft() {
		swipe(Direction.LEFT);
		return gridOfTiles.get();
	}

	private void swipe(Direction direction) {
		gridOfTiles.takeSnapshot();

		gridOfTiles.migrate(direction);

		movesCount++;

		checkForVictory(gridOfTiles.getCurrentMax());

		try {
			gridOfTiles.newTile();
		} catch (IllegalArgumentException e) { //no empty spaces.
			checkGameOver();
		}
	}

	@Override
	public Tile[][] undo() {
		gridOfTiles.useSnapshotTiles();
		return gridOfTiles.get();
	}

	@Override
	public void reset() {
		if (isStarted) {
			gridOfTiles.clear();

			isVictory = false;
			isGameOver = false;
			isStarted = false;

			movesCount = 0;
		}
	}

	@Override
	public boolean isGameOver() {
		return isGameOver;
	}

	@Override
	public boolean isItVictory() {
		return isVictory;
	}


	private void checkGameOver() {
		if (!gridOfTiles.isMovingPossible())
			throw new GameOverException("You lost");
	}

	private void checkForVictory(int value) { //TODO
		isVictory = (value >= maxValue);
	}

	public int getMovesCount() {
		return movesCount;
	}
}

File: src/mhashim6/game1024/GameImpl.kt
package mhashim6.game1024

import mhashim6.game1024.exceptions.GameOverException
import mhashim6.game1024.Direction.*


class GameImpl(power: Int, gridSize: Int) : Game {
	private val gridOfTiles: GridOfTiles = GridOfTiles(gridSize)

	private val maxValue: Int = Math.pow(2.0, power.toDouble()).toInt() //11 -> 2048

	private var movesCount: Int = 0
	override fun movesCount() = movesCount

	private var isStarted: Boolean = false

	override fun start(): Array<Array<Tile?>> {
		if (isStarted)
			throw RuntimeException("The game is already on") //TODO better
		isStarted = true

		gridOfTiles.newTile()
		gridOfTiles.newTile()

		return gridOfTiles.copy()
	}

	override fun swipeUp(): Array<Array<Tile?>> {
		swipe(UP)
		return gridOfTiles.copy()
	}

	override fun swipeDown(): Array<Array<Tile?>> {
		swipe(DOWN)
		return gridOfTiles.copy()
	}

	override fun swipeRight(): Array<Array<Tile?>> {
		swipe(RIGHT)
		return gridOfTiles.copy()
	}

	override fun swipeLeft(): Array<Array<Tile?>> {
		swipe(LEFT)
		return gridOfTiles.copy()
	}

	private fun swipe(direction: Direction) {
		gridOfTiles.takeSnapshot()
		gridOfTiles.migrate(direction)

		movesCount++

		updateState()
		gridOfTiles.newTile()
	}

	private fun updateState() {
		movesCount++

		if (isVictory(gridOfTiles.currentMax))
			throw GameOverException("You Win!")

		if (!gridOfTiles.isMovingPossible)
			throw GameOverException("You Lose!")
	}

	override fun undo(): Array<Array<Tile?>> {
		gridOfTiles.useSnapshot()
		return gridOfTiles.copy()
	}

	override fun reset() {
		if (isStarted) {
			gridOfTiles.clear()

			isStarted = false

			movesCount = 0
		}
	}

	private fun isVictory(value: Int) = value >= maxValue
}
